sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, size_t srcStride);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, size_t destStride);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, size_t srcStride, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, size_t destStride, sycl::sub_group_mask mask);
+} // namespace sycl::ext::oneapi
+```
+
+## Change to `device_event::wait` function
+`device_event::wait` will be modified to take the group that will be synchonised in the wait.
+This will be valid for the same groups as `async_group_copy` i.e. `group` or `sub_group`.
+An overload with a `sub_group_mask` parameter will be required for `sub_group`.
+
+```c++
+namespace sycl {
+  class device_event {
+  ...
+  public:
+    template <typename Group>
+    void ext_oneapi_wait(Group group) noexcept;
+    void ext_oneapi_wait(sycl::sub_group group, sycl::sub_group_mask mask) noexcept;
+  ...
+  };
+}  // namespace sycl
+```
+
+## Open Questions
+
+* Do the `sub_group_mask` overloads of functions also need the `sub_group` class as an argument, since this is implicit.
+* How to handle the bitmask partition not dividing the sub_group size evenly?
+* What to do if the bitmask sub_group partition doesn't divide the size of the nd_range evenly? Potentially UB.
+* Should async_group_copy destination and source be flipped to match sycl::handler::copy and sycl::handler::memcopy.
+* Do we need to introduce an instance of sub_group_mask with the number of bits set as a compile-time known value to allow optimisation?
+
+## Alternate Implementation
```

Comment:
A few of us at Intel had been working on an idea similar to this.  I've posted it as a draft PR in https://github.com/intel/llvm/pull/4904 -- I'd appreciate your thoughts.  Tagging @AerialMantis, @JackAKirk as well.

Reply:
I read #4904, nice summary.  From my current understanding it reads as though the two proposals map exactly if you map explicit_mask to `ext_oneapi_partition_sub_group` and implicit_mask to `ext_oneapi_active_sub_group_items`.

Reply:
I think you may be right.  I was initially concerned that the definitions of "active work-items" were different, but re-reading things this morning I've convinced myself that they both allow for implementation-defined convergence guarantees (e.g. for architectures like Volta).

Reply:
We are currently working on the implementation that uses both types of masks with the group collective functions: async_group_copy, wait, sync etc.  Hopefully this will be up quite soon which should show a concrete example of how the "explicit" and "implicit" masks can be used with the CUDA backend.